Remove 0-size and variable size array members

yuyichao opened the pull request:
For some reason, the `llvm-config` from LLVM 3.7 adds `-pedantic` to cflags and cxxflags. This causes a huge amount of compiler warnings (mainly because of `JL_DATA_TYPE`). I couldn't find out why did LLVM starts doing this or how to workaround it in general so I decided to fix most of the warnings....

This PR doesn't completely fix all the `-Wpedantic` warnings (the ones that are not fixed is anonymous structs which we use for bitsfields in a union. This turns out to be valid C11 but not valid C++ yet...).

This could break embedding. In particular,
1. `jl_value_t` is only a declaration now. Direct use of `jl_value_t` instead of `jl_value_t*` shouldn't be necessary.
2. Types with variable array members do no have such member anymore. The address of the variable length array is calculated directly with pointer arithmetics instead. Notes about how each of them should be updated,
   1. `jl_value_t::fieldptr`
      
      We already have `jl_data_ptr` and this macro should be used instead.
   2. `jl_sym_t::data`
      
      Use `jl_symbol_name`, which is also the name of the exported function.
   3. `jl_taggedvalue_t::value`
      
      (not actually a flexible array member but was serving a similar purpose). I doubt any embedding code manipulate `jl_taggedvalue_t` directly but `jl_valueof` is provided as a convenient way to access the value pointer.
   4. `jl_svec_t::data`
      
      We already have `jl_svec_data` and this macro should be used instead.
   5. `jl_datatype_t::fields`
      
      This shouldn't be accessed directly since https://github.com/JuliaLang/julia/pull/13147. `jl_datatype_fields` macro is also added to get the address of the fields in the rare case it is needed.
3. If any code was relying on `JL_DATA_TYPE` in a non-gc allocated `struct` to provide necessary alignment, it needs to find another way to do it. (define its own zero length array member, use the alignment attributes etc.)
4. The presence of `jl_valueof` macro and function could be used as a compile time and runtime check to detect this change.

@vtjnash Ref https://github.com/JuliaLang/julia/pull/13787
@tkelman This should make the code more MSVC compatible but it's also possible that I break something without realizing it....

yuyichao committed:
Remove 0-size and variable size array members from public headers.

tkelman commented:
built fine on msvc straight away, nicely done.

vtjnash commented on src/codegen.cpp:
```diff
@@ -709,10 +709,11 @@ typedef struct {
         int64_t isref;
         Function *f;
-    } data[];
+    } data[1];
```
I'm pretty sure this change isn't ISO-conforming C. The old version should have been more correct.

yuyichao replied:
Why is this not ISO-conforming about this?

The warning I get is
```
yuyichao% LANG=C g++ -pedantic -x c++ -Wall -Wextra - -c -o /dev/null <<EOF
typedef struct {
        int a;
        int b[];
} A;
EOF
<stdin>:3:8: warning: ISO C++ forbids zero-size array 'b' [-Wpedantic]
```

yuyichao replied:
OK so I find [this](http://stackoverflow.com/questions/3711233/is-the-struct-hack-technically-undefined-behavior/) stackoverflow thread. And it seems that there basically isn't a ISO C++ conforming way to do this.

* The flexible member is invalid C++ syntax and causes warning under `-pedantic`.

* The single element array could technically be UB but `It's not clear if it's legal or portable, but it is rather popular.` and `does seem to work under all known implementations`.... (and also issue no warnings.)

vtjnash replied:
> does seem to work under all known implementations 

http://stackoverflow.com/a/10500898

vtjnash replied:
perhaps we should just switch this struct to a `std::unordered_map< uint64_t, Function* >`

Keno replied:
Doing the [1] thing can now give you runtime errors on OS X, cf  https://github.com/JuliaLang/julia/pull/13845

yuyichao replied:
> perhaps we should just switch this struct to a `std::unordered_map< uint64_t, Function* >`

We don't require `c++11` in general yet but it could be `std::map`? Looks like it currently is simply a `std::vector`?

vtjnash replied:
`std::vector< std::pair< int64_t, Function* > >` would also work. it's expected to be a very small set, so actually it's best not to use `unordered_map`, which may significantly over-allocate.

vtjnash commented on src/codegen.cpp:
```diff
@@ -5482,8 +5483,8 @@ static void init_julia_llvm_env(Module *m)
         71, // At the time of this writing. Not sure if it's worth it to keep this in sync
-        sizeof(jl_value_t)*8,
-        __alignof__(jl_value_t)*8,
+        0 * 8, // sizeof(jl_value_t) * 8,
```
@Keno: I think we've had a conversation about before, but I don't remember the result. Is DWARF going to dislike a declared size of 0 bits?

yuyichao replied:
This is not new and has always been `0`. (At least on 0.4).

Keno replied:
Should be fine with DWARF.

vtjnash commented on src/jlapi.c:
```diff
@@ -310,6 +310,13 @@ DLLEXPORT jl_taggedvalue_t *jl_astaggedvalue(jl_value_t *v)
+DLLEXPORT jl_value_t *jl_valueof(jl_taggedvalue_t *v)
+{
+    return jl_valueof__MACRO(v);
+}
```
these would be easier to use if we just directly shadow the names (e.g. it might avoid the gdb `define macro jl_typeof jl_typeof` dance). I didn't do this initially because I was concerned some C parser out there would not handle these correctly:
```diff
diff --git a/src/jlapi.c b/src/jlapi.c
index a654d74..051d40f 100644
--- a/src/jlapi.c
+++ b/src/jlapi.c
@@ -304,16 +304,14 @@ DLLEXPORT const char *jl_git_commit()
 }
 
 // Create function versions of some useful macros
-#undef jl_astaggedvalue
-DLLEXPORT jl_taggedvalue_t *jl_astaggedvalue(jl_value_t *v)
+DLLEXPORT jl_taggedvalue_t *(jl_astaggedvalue)(jl_value_t *v)
 {
-    return jl_astaggedvalue__MACRO(v);
+    return jl_astaggedvalue(v);
 }
 
-#undef jl_typeof
-DLLEXPORT jl_value_t *jl_typeof(jl_value_t *v)
+DLLEXPORT jl_value_t *(jl_typeof)(jl_value_t *v)
 {
-    return jl_typeof__MACRO(v);
+    return jl_typeof(v);
 }
 
 #ifdef __cplusplus
diff --git a/src/julia.h b/src/julia.h
index 1d033e3..694eea5 100644
--- a/src/julia.h
+++ b/src/julia.h
@@ -172,10 +172,8 @@ typedef struct {
     jl_value_t value;
 } jl_taggedvalue_t;
 
-#define jl_astaggedvalue__MACRO(v) container_of((v),jl_taggedvalue_t,value)
-#define jl_typeof__MACRO(v) ((jl_value_t*)(jl_astaggedvalue__MACRO(v)->type_bits&~(uintptr_t)15))
-#define jl_astaggedvalue jl_astaggedvalue__MACRO
-#define jl_typeof jl_typeof__MACRO
+#define jl_astaggedvalue(v) container_of((v),jl_taggedvalue_t,value)
+#define jl_typeof(v) ((jl_value_t*)(jl_astaggedvalue(v)->type_bits&~(uintptr_t)15))
 static inline void jl_set_typeof(void *v, void *t)
 {
     jl_taggedvalue_t *tag = jl_astaggedvalue(v);
```

yuyichao replied:
I was wondering if it is written this way because of some wierd pre-processor implementations.....

vtjnash replied:
> So are you sure this is fine with all pre-processors now?

Yes. With further research, I realized this works because it hides the essential parentheses from the function call macro, not just some random implementation-specific behavior.

vtjnash commented on src/julia.h:
```diff
@@ -169,12 +166,15 @@ typedef struct {
     };
-    jl_value_t value;
+    // variable length
+    void *value[1];
```
this appears to change the size of `jl_taggedvalue_t`, which isn't entirely optimal for gdb. this may be the best we can do, but so long as we are doing this, perhaps we can make it more even useful:
```
typedef union {
    uint8_t u8;
    int8_t i8;
    ...
    struct _jl_lambda_info_t linfo;
    struct _jl_expr_t expr;
    struct _jl_array_t array;
    struct _jl_svec_t svec; // pretty sure this one won't be viable, because it is variable length (same for jl_datatype_t)
} jl_value_t;
```
then a `jl_value_t*` (such as is found in `jl_breakpoint`) can be "cast" in gdb to a lambda-info (for example) by writing `v->linfo`

vtjnash commented on src/julia.h:
```diff
@@ -188,7 +188,8 @@ typedef struct _jl_sym_t {
     uptrint_t hash;    // precomputed hash value
-    JL_ATTRIBUTE_ALIGN_PTRSIZE(char name[]);
+    // variable length
+    JL_ATTRIBUTE_ALIGN_PTRSIZE(char name[1]);
```
for these, maybe we should just eliminate these convenience members and use macros instead to point to these elements as `#define jl_sym_name(sym) ((char*)(((jl_sym_t*)(sym)) + 1))` (it will be unfortunate to lose the ability to look at `p/s name->name` in gdb though)

yuyichao replied:
GDB problem can be solved with `const char *(jl_sym_name)(jl_sym_t *sym) {return jl_sym_name(sym);}` ?

vtjnash replied:
yep, same as all those other macros. fwiw, i'm not certain if it should be `jl_sym_t` or `jl_value_t` there, since lldb and C care about the exact pointer type, usually the src code sticks with jl_value_t to reduce the amount of casting.

yuyichao commented:
I updated the PR to not use `[1]` and calculate the addresses directly instead. I've implemented most of what @vtjnash suggested except,

1. Use `std::vector` or `std::map` for `cFunctionList`. It's requires slightly more changes to implement and introduce one more indirection since we can only store a pointer in the C structure.

2. Define `jl_value_t` as a `union`. It's largely independent of this PR.

The `static inline` version of `jl_symbol_name_` is to provide better type check. `jl_sym_t::name` is used a lot in the code and usually in the form of `...->name->name` (sometimes `...->name->name->name`) and is really likely to get wrong if there's not a strong type check.

vtjnash commented:
lgtm

tkelman commented:
6db03d00e2888bec72a3a8b025486defc126317a looks like it also works on msvc. Do any of the embedding docs or examples need updating for this?

yuyichao commented:
> Do any of the embedding docs or examples need updating for this?

Good question. All the changes are API level details (i.e. shouldn't break ABI) and it seems that none of these are covered by the embedding doc. (We don't have doc for the various `jl_*_t` types).

yuyichao commented:
Rebased to resolve conflict with https://github.com/JuliaLang/julia/pull/13787 .... I guess I shouldn't have silently included the fix for the variable not used warning when assert is off....

vtjnash commented:
Oops, i guess at least one of us shouldn't have silently added that fix. We very nearly came up with the same patch too :/

yuyichao commented:
Rebased and add a short comment about `jl_symbol_name_`. Will merge tomorrow if there's no other issues.
